src/platform/datapath_raw_socket.c

```diff
@@ -307,27 +312,115 @@ CxPlatDpRawParseEthernet(
         return;
     }
 
-    if (Ethernet->Type == 0x0008) { // IPv4
+    uint16_t EthernetType = Ethernet->Type;
+    if (EthernetType == ETHERNET_TYPE_IPV4) {
         CxPlatDpRawParseIPv4(Datapath, Packet, (IPV4_HEADER*)Ethernet->Data, Length);
-    } else if (Ethernet->Type == 0xDD86) { // IPv6
+    } else if (EthernetType == ETHERNET_TYPE_IPV6) {
         CxPlatDpRawParseIPv6(Datapath, Packet, (IPV6_HEADER*)Ethernet->Data, Length);
+    } else {
+        QuicTraceEvent(
+            LibraryErrorStatus,
+            "[ lib] ERROR, %u, %s.",
+            EthernetType,
+            "unacceptable Ethernet type");
     }
 }
 
+_IRQL_requires_max_(DISPATCH_LEVEL)
+HEADER_BACKFILL
+CxPlatDpRawCalculateHeaderBackFill(
+    _In_ QUIC_ADDRESS_FAMILY Family
+    )
+{
+    HEADER_BACKFILL HeaderBackFill;
+    HeaderBackFill.TransportLayer = sizeof(UDP_HEADER);
+    HeaderBackFill.NetworkLayer =
+        Family == QUIC_ADDRESS_FAMILY_INET ? sizeof(IPV4_HEADER) : sizeof(IPV6_HEADER);
+    HeaderBackFill.LinkLayer = sizeof(ETHERNET_HEADER);
+    HeaderBackFill.AllLayer =
+        HeaderBackFill.TransportLayer + HeaderBackFill.NetworkLayer + HeaderBackFill.LinkLayer;
+    return HeaderBackFill;
+}
+
 _IRQL_requires_max_(DISPATCH_LEVEL)
 void
 CxPlatFramingWriteHeaders(
     _In_ const CXPLAT_SOCKET* Socket,
     _In_ const QUIC_ADDR* LocalAddress,
     _In_ const QUIC_ADDR* RemoteAddress,
-    _Inout_ QUIC_BUFFER* Buffer
+    _Inout_ QUIC_BUFFER* Buffer,
+    _In_ QUIC_ADDRESS_FAMILY Family
     )
 {
     UDP_HEADER* UDP = (UDP_HEADER*)(Buffer->Buffer - sizeof(UDP_HEADER));
-    IPV4_HEADER* IP = (IPV4_HEADER*)(((uint8_t*)UDP) - sizeof(IPV4_HEADER));
-    ETHERNET_HEADER* Ethernet = (ETHERNET_HEADER*)(((uint8_t*)IP) - sizeof(ETHERNET_HEADER));
+    ETHERNET_HEADER* Ethernet;
```

Comment:
Might there be a slight cache locality improvement to write the header from front (ethernet) to back (udp) order?

Reply:
Can you explain why? Even if we try to write eth header first, the cache doesn't necessarily start with the address of the eth header. Whether all headers are in the same line is not determined by which byte we first try to access.

Reply:
Just wondering.